api/src/main/java/com/epam/pipeline/security/saml/SAMLUserDetailsServiceImpl.java

```diff
@@ -77,39 +84,54 @@
 
     @Override
     public UserContext loadUserBySAML(SAMLCredential credential) {
-        String userName = credential.getNameID().getValue().toUpperCase();
-        List<String> groups = readAuthorities(credential);
-        Map<String, String> attributes = readAttributes(credential);
-
-        PipelineUser loadedUser = userManager.loadUserByName(userName);
-        if (loadedUser == null) {
-            String message = messageHelper.getMessage(MessageConstants.ERROR_USER_NAME_NOT_FOUND, userName);
-            checkAbilityToCreate(userName, groups);
-            LOGGER.debug(message);
-            List<Long> roles = roleManager.getDefaultRolesIds();
-            PipelineUser createdUser = userManager.createUser(userName,
-                    roles, groups, attributes, null);
-            LOGGER.debug("Created user {} with groups {}", userName, groups);
-            validateGroupsBlockingStatus(createdUser.getAuthorities(), userName);
-            UserContext userContext = new UserContext(createdUser.getId(), userName);
-            userContext.setGroups(createdUser.getGroups());
-            userContext.setRoles(createdUser.getRoles());
-            return userContext;
+        final String userName = credential.getNameID().getValue().toUpperCase();
+        final List<String> groups = readAuthorities(credential);
+        final Map<String, String> attributes = readAttributes(credential);
+        final UserContext userContext = Optional.ofNullable(userManager.loadUserByName(userName))
+            .map(loadedUser -> processRegisteredUser(userName, groups, attributes, loadedUser))
+            .orElseGet(() -> processNewUser(userName, groups, attributes));
+        if (hasBlockedStatusAttribute(credential)) {
+            userManager.updateUserBlockingStatus(userContext.getUserId(), true);
+            throwUserIsBlocked(userName);
+        }
+        return userContext;
+    }
+
+    private UserContext processNewUser(final String userName, final List<String> groups,
+                                       final Map<String, String> attributes) {
+        final String message = messageHelper.getMessage(MessageConstants.ERROR_USER_NAME_NOT_FOUND, userName);
+        checkAbilityToCreate(userName, groups);
+        LOGGER.debug(message);
+        final List<Long> roles = roleManager.getDefaultRolesIds();
+        final PipelineUser createdUser = userManager.createUser(userName,
+                                                                roles, groups, attributes, null);
+        LOGGER.debug("Created user {} with groups {}", userName, groups);
+        final UserContext userContext = new UserContext(createdUser.getId(), userName);
+        userContext.setGroups(createdUser.getGroups());
+        userContext.setRoles(createdUser.getRoles());
+        validateGroupsBlockingStatus(createdUser.getAuthorities(), userName);
```

Comment:
Since this method is executed for both: new and existing users, we can move it common code in `loadUserBySAML` method. Use `UserContext.getAuthorities` method for filtering. Good point - in this class authorities doesn't contain username, so we don't need to filter it out in `validateGroupsBlockingStatus` method.